bindings/rust/s2n-tls/src/connection.rs

```diff
@@ -32,6 +32,7 @@ macro_rules! static_const_str {
     };
 }
 
+#[derive(PartialEq)]
```

Comment:
Why not go all the way and derive `Eq`?

Is it possible to make the derive explicitly only for test cases?
```
#[cfg_attr(test,derive(PartialEq))]
```
Does the above work? -- or is there something similar that would work?

Reply:
> Why not go all the way and derive Eq?

I had originally implemented it as `#[derive(Eq)]`, but the compiler wouldn't allow it and instead recommended that I implement `PartialEq`. Full error plus recommendation for reference:
```
error[E0277]: can't compare `Connection` with `Connection`
   --> s2n-tls/src/connection.rs:35:10
    |
35  | #[derive(Eq)]
    |          ^^ no implementation for `Connection == Connection`
    |
    = help: the trait `PartialEq` is not implemented for `Connection`
note: required by a bound in `std::cmp::Eq`
   --> /Users/jonahmil/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/cmp.rs:292:15
    |
292 | pub trait Eq: PartialEq<Self> {
    |               ^^^^^^^^^^^^^^^ required by this bound in `std::cmp::Eq`
    = note: this error originates in the derive macro `Eq` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `Connection` with `#[derive(PartialEq)]`
    |
36  | #[derive(PartialEq)]
    |
```
With that said, I played with it a bit and this error can also be fixed by annotating the Connection struct with `#[derive(PartialEq, Eq)]` instead, which is a much better implementation as it allows for self-comparisons (`a == a`). I'm happy to post a revision with both `PartialEq` and `Eq` implemented.
If you're curious, read [here](https://users.rust-lang.org/t/what-is-the-difference-between-eq-and-partialeq/15751) for further detail.

> Is it possible to make the derive explicitly only for test cases?

I'm sure it's possible, but it doesn't make sense to do so. It's reasonable to expect that some future developer using `s2n-tls` would want to compare two `Connection` instances, so there's no reason to limit the implementation to unit tests only.

If I as a consumer of `s2n-tls` was to dig into the code to see whether it was possible to compare `Connection`s, I'd be irked to discover that someone had found a use case for `Eq` but implemented it so that it worked for internal tests only. As long as there's no security risk, there's no reason not to implement for general use and anticipate the needs of future users.

Reply:
I'm gonna let someone else weigh in on weather or not comparing connections makes sense. I'd ask if there is some other notion of equality we should be checking (as opposed to mere pointer equality). But if mere pointer equality makes sense then I'd suggest deriving Eq as well -- especially since it is implemented for `NonNull<T>`

Reply:
I think I would prefer not to implement `PartialEq` for now.

Reply:
Can I ask why not? Is there risk associated with doing so?

Reply:
I think the thing here is what does it mean for two connections to be equal? is it pointer equivalence (that's what's being implemented here at least)? but it think it isn't unreasonable to expect a structural equivalence, as in do the host names that we're connecting to match? like if i do

```rust
assert_eq!(Box::new(123), Box::new(123))
```

This will pass because we're doing a _value_ comparison even though they're different pointers. And one of the downsides to this trait impl is we can't document what we mean when we implement it.

Reply:
Makes sense. I'm not married to implementing `PartialEq` or `Eq`, as your suggested implementation works just as well.

With that said, could we hypothetically implement `PartialEq` and/or `Eq` ourselves to perform a value comparison rather than a reference comparison, or are those traits somehow tied to referential comparison?

Reply:
yeah you can make the `PartialEq` do whatever you want. but it gets more into what does it mean for two connections to be "value equal" and I don't think we have a good answer. or at least we don't have a compelling use case to actually define it IMO.